contracts/EtomicSwap.sol

```diff
@@ -115,25 +200,194 @@ contract EtomicSwap {
 
         payments[_id].state = PaymentState.SenderRefunded;
 
         if (_tokenAddress == address(0)) {
-            msg.sender.transfer(_amount);
+            payable(msg.sender).transfer(_amount);
         } else {
             IERC20 token = IERC20(_tokenAddress);
             require(token.transfer(msg.sender, _amount));
         }
 
         emit SenderRefunded(_id);
     }
+
+    function senderRefundErc721(
+        bytes32 _id,
+        bytes20 _paymentHash,
+        address _tokenAddress,
+        uint256 _tokenId,
+        address _receiver
+    ) external {
+        require(payments[_id].state == PaymentState.PaymentSent);
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                _receiver,
+                msg.sender,
+                _paymentHash,
+                _tokenAddress,
+                _tokenId
+            )
+        );
+
+        require(
+            paymentHash == payments[_id].paymentHash &&
+            block.timestamp >= payments[_id].lockTime
+        );
+        payments[_id].state = PaymentState.SenderRefunded;
+
+        IERC721 token = IERC721(_tokenAddress);
+        token.safeTransferFrom(address(this), msg.sender, _tokenId);
+
+        emit SenderRefunded(_id);
+    }
+
+    function senderRefundErc1155(
+        bytes32 _id,
+        uint256 _amount,
+        bytes20 _paymentHash,
+        address _tokenAddress,
+        uint256 _tokenId,
+        address _receiver
+    ) external {
+        require(payments[_id].state == PaymentState.PaymentSent);
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                _receiver,
+                msg.sender,
+                _paymentHash,
+                _tokenAddress,
+                _tokenId,
+                _amount
+            )
+        );
+
+        require(
+            paymentHash == payments[_id].paymentHash &&
+            block.timestamp >= payments[_id].lockTime
+        );
+        payments[_id].state = PaymentState.SenderRefunded;
+
+        IERC1155 token = IERC1155(_tokenAddress);
+        token.safeTransferFrom(
+            address(this),
+            msg.sender,
+            _tokenId,
+            _amount,
+            ""
+        );
+
+        emit SenderRefunded(_id);
+    }
+
+    function onERC1155Received(
+        address, /* operator */
+        address from,
+        uint256 tokenId,
+        uint256 value,
+        bytes calldata data
+    ) external override returns (bytes4) {
+        // Decode the data to extract HTLC parameters
+        (
+            bytes32 id,
+            address receiver,
+            address tokenAddress,
+            bytes20 secretHash,
+            uint64 lockTime
+        ) = abi.decode(data, (bytes32, address, address, bytes20, uint64));
+
+        require(
+            receiver != address(0) &&
+            tokenAddress != address(0) &&
+            value > 0 &&
+            payments[id].state == PaymentState.Uninitialized
+        );
```

Comment:
I think it's worth adding the following conditions:
1. `msg.sender == tokenAddress` to ensure that the function is called by the token contract. Maybe we can even reduce the data size and use `msg.sender` instead of `tokenAddress`. I'm not sure regarding the possible security risks of this, please check that it's safe :slightly_smiling_face: 
1. `tx.origin == from && tx.origin == operator` to ensure that token transfer is made by user account itself.
1. Might be also worth checking that receiver is not a smart contract address.

Reply:
Done ([comment](https://github.com/KomodoPlatform/etomic-swap/pull/2#discussion_r1410332455))